src/opts.c

```diff
@@ -883,17 +883,190 @@ natsOptions_SetMaxPendingMsgs(natsOptions *opts, int maxPending)
+            }
+            else
+            {
+                prev->next = this->next;
+            }
+            
+            return;
+        }
+    }
+}
+
+void
+natsOptions_freeCallbackList(nats_CallbackList *cb)
+{
+    nats_CallbackList *next = NULL;
+
+    while (cb != NULL)
+    {
+        next = cb->next;
+        NATS_FREE(cb);
+        cb = next;
+    }
+}
+
+natsStatus
+natsOptions_cloneCallbackList(nats_CallbackList **clone, nats_CallbackList *this)
+{
+    nats_CallbackList *prev = NULL;
+    nats_CallbackList *head = NULL;
+    nats_CallbackList *tail = NULL;
+
+    for (; this != NULL; this = this->next)
+    {
+        tail = NATS_CALLOC(1, sizeof(nats_CallbackList));
+        if (tail == NULL)
+        {
+            natsOptions_freeCallbackList(head);
+            return nats_setDefaultError(NATS_NO_MEMORY);
+        }
+        *tail = *this;
+        tail->next = NULL;
+
+        if (prev == NULL)
+        {
+            head = tail;
+        }
+        else
+        {
+            prev->next = tail;
+        }
+        prev = tail;
+    }
+
+    *clone = head;
+    return NATS_OK;
+}
+
+natsStatus
+natsOptions_addConnectionClosedCallback(natsOptions *opts, natsConnectionHandler f, void *closure)
+{
+    natsStatus          s           = NATS_OK;
+    nats_CallbackList   *cb         = NULL;
+    nats_CallbackList   *replaced   = NULL;
+
+    cb = NATS_CALLOC(1, sizeof(nats_CallbackList));
+    if (cb == NULL)
+        return nats_setDefaultError(NATS_NO_MEMORY);
+    cb->type = CALLBACK_TYPE_CONN;
+    cb->f.conn = f;
+    cb->closure = closure;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    // unlink if already in the list.
+    natsOptions_unlinkCallback(&replaced, &opts->closedCb, (void(*)(void))f, closure);
+
+    // add at the head.
+    cb->next = opts->closedCb;
+    opts->closedCb = cb;
+
+    UNLOCK_OPTS(opts);
+
+    NATS_FREE(replaced);
+
+    return s;
+}
+
+natsStatus
+natsOptions_addErrorCallback(natsOptions *opts, natsErrHandler f, void *closure)
+{
+    natsStatus          s           = NATS_OK;
+    nats_CallbackList   *cb         = NULL;
+    nats_CallbackList   *replaced   = NULL;
+
+    cb = NATS_CALLOC(1, sizeof(nats_CallbackList));
+    if (cb == NULL)
+        return nats_setDefaultError(NATS_NO_MEMORY);
+    cb->type = CALLBACK_TYPE_ERROR;
+    cb->f.err = f;
+    cb->closure = closure;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    // unlink if already in the list.
+    natsOptions_unlinkCallback(&replaced, &opts->asyncErrCb, (void(*)(void))f, closure);
+
+    // add at the head.
+    cb->next = opts->asyncErrCb;
+    opts->asyncErrCb = cb;
+
+    UNLOCK_OPTS(opts);
+
+    return s;
+}
+
+void natsOptions_removeConnectionClosedCallback(natsOptions *opts, natsConnectionHandler f, void *closure)
+{
+    nats_CallbackList *removed = NULL;
+
+    natsMutex_Lock(opts->mu);
+
+    natsOptions_unlinkCallback(&removed, &opts->closedCb, (void(*)(void))f, closure);
+
+    UNLOCK_OPTS(opts);
```

Comment:
nit: but since you used explicitly `natsMutex_Lock(opts->mu)` above, I would use `natsMutex_Unlock()` here, or use the macro to lock - just for consistency sake.